src/Api/Auth/Controllers/TwoFactorController.cs

```diff
@@ -285,30 +295,46 @@ public async Task SendEmail([FromBody] TwoFactorEmailRequestModel model)
 
     [AllowAnonymous]
     [HttpPost("send-email-login")]
-    public async Task SendEmailLogin([FromBody] TwoFactorEmailRequestModel model)
+    public async Task SendEmailLoginAsync([FromBody] TwoFactorEmailRequestModel requestModel)
     {
-        var user = await _userManager.FindByEmailAsync(model.Email.ToLowerInvariant());
+        var user = await _userManager.FindByEmailAsync(requestModel.Email.ToLowerInvariant());
+
         if (user != null)
         {
             // check if 2FA email is from passwordless
-            if (!string.IsNullOrEmpty(model.AuthRequestAccessCode))
+            if (!string.IsNullOrEmpty(requestModel.AuthRequestAccessCode))
             {
                 if (await _verifyAuthRequestCommand
-                        .VerifyAuthRequestAsync(new Guid(model.AuthRequestId), model.AuthRequestAccessCode))
+                        .VerifyAuthRequestAsync(new Guid(requestModel.AuthRequestId),
+                            requestModel.AuthRequestAccessCode))
                 {
                     await _userService.SendTwoFactorEmailAsync(user);
                     return;
                 }
             }
-            else if (await _userService.VerifySecretAsync(user, model.Secret))
+            else if (!string.IsNullOrEmpty(requestModel.SsoEmail2FaSessionToken))
+            {
+                if (this.ValidateSsoEmail2FaToken(requestModel.SsoEmail2FaSessionToken, user))
+                {
+                    await _userService.SendTwoFactorEmailAsync(user);
+                    return;
+                }
+                else
+                {
+                    this.ThrowDelayedBadRequestExceptionAsync(
+                        "Cannot send two-factor email: a valid, non-expired SSO Email 2FA Session token is required to send 2FA emails.",
+                        2000);
+                }
+            }
+            else if (await _userService.VerifySecretAsync(user, requestModel.Secret))
             {
                 await _userService.SendTwoFactorEmailAsync(user);
                 return;
             }
         }
 
-        await Task.Delay(2000);
-        throw new BadRequestException("Cannot send two-factor email.");
+        this.ThrowDelayedBadRequestExceptionAsync(
```

Comment:
Missing an `await`?

Reply:
![image](https://user-images.githubusercontent.com/116684653/234654991-e379a50e-7cdd-4647-9297-cfa657824d77.png)

I thought so too initially, but it's not awaitable b/c it doesn't return anything and just throws.

Reply:
Ah, thank you for the explanation. I did not know that. I've updated it to be `async Task` after some research.